### PR TITLE
New functionality from charmcraft (CRAFT-706).

### PR DESCRIPTION
**Pull request by facundobatista**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_cli/dispatcher.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    It also must/can override some methods for the proper command behaviour (see each
+    method's docstring).
+
+    The subclass must be declared in the corresponding section of main.COMMAND_GROUPS,
+    and will receive and store this group on instantiation (if overriding `__init__`, the
+    subclass must pass it through upwards).
+    """
+
+    common = False
+    needs_config = False
```
Comment (facundobatista) [inline craft_cli/dispatcher.py:90]: I think I will remove this attribute from here.

If the commands needs a config or not is something very charmcraft specific, and it actually cannot be automatically verified here (charmcraft needs to check it!).

File: craft_cli/helptexts.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+import textwrap
+from operator import attrgetter
+from typing import TYPE_CHECKING, List, Tuple
+
+if TYPE_CHECKING:
+    from craft_cli.dispatcher import BaseCommand, CommandGroup
+
+
+# max columns used in the terminal
+TERMINAL_WIDTH = 72
```
Comment (cmatsuoka) [inline craft_cli/helptexts.py:27]: Why do we use 72 instead of, say, 78?

File: Makefile [hunk 3]
```diff
@@ -91,7 +91,7 @@ test-pydocstyle:
 
 .PHONY: test-pylint
 test-pylint:
-	pylint craft_cli
+	pylint craft_cli --disable=cyclic-import
```
Comment (cmatsuoka) [inline Makefile:94]: Why is this needed? Is it reporting a false positive?

**Review by cmatsuoka — APPROVED** [review]
Since this was extracted from Charmcraft I assume it will provide all the functionality needed by this tool. We'll see how generic it is when integrating with other craft tools, but generally it LGTM. We may need to rethink the legacy fallback mechanism implementation (as previously discussed) because the spike just parsed arguments by hand.

File: craft_cli/dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
+# Copyright 2021 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_cli/dispatcher.py:1]: Updated copyright years here and in all the files.

File: craft_cli/helptexts.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
+# Copyright 2021 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_cli/helptexts.py:1]: Yes.

File: craft_cli/helptexts.py (hunk 2 above)
Comment (facundobatista) [reply craft_cli/helptexts.py:27]: I wanted to be conservative here. Note that this module does not present the texts to the user, just builds them and returns them. Maybe the caller would want to indent them one or two columns... so I decided not to use "80", and choose what PEP8 recommends for docstrings, which is "72".

But I'm to change it if you have reasons.

File: Makefile (hunk 3 above)
Comment (facundobatista) [reply Makefile:94]: Yes. Note that import order in `craft_cli/__init__.py` is defined to avoid the cyclic import. But pylint falls in "the trap" anyway...

File: craft_cli/helptexts.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_cli/helptexts.py:27]: I'm good with 72, I was just curious about this exact value.

File: Makefile (hunk 3 above)
Comment (sergiusens) [reply Makefile:94]: If you need to toggle a check, do it here instead https://github.com/canonical/craft-cli/blob/main/.pylintrc#L72

**Review by sergiusens — CHANGES_REQUESTED** [review]
just wanted to flag that the check should be blocked in configuration for all IDEs to benefit

File: Makefile (hunk 3 above)
Comment (facundobatista) [reply Makefile]: ack, improved.

**Review by sergiusens — APPROVED** [review]
